Jenkinsfile

```diff
@@ -0,0 +1,399 @@
+        /**
+          Run unit tests and report junit results.
+        */
+        stage('Unit Test') {
+          agent { label 'linux && immutable' }
+          options { skipDefaultCheckout() }
+          environment {
+            PATH = "${env.PATH}:${env.WORKSPACE}/bin"
+            HOME = "${env.WORKSPACE}"
+            GOPATH = "${env.WORKSPACE}"
+          }
+          when { 
+            beforeAgent true
+            environment name: 'test_ci', value: 'true' 
+          }
+          steps {
+            withEnvWrapper() {
+              unstash 'source'
+              dir("${BASE_DIR}"){
+                sh './script/jenkins/unit-test.sh'
+              }
+            }
+          }
+          post { 
+            always {
+              junit(allowEmptyResults: true, 
+                keepLongStdio: true, 
+                testResults: "${BASE_DIR}/build/junit-*.xml")
+            }
+          }
+        }
+        /**
+        Runs System and Environment Tests, then generate coverage and unit test reports.
+        Finally archive the results.
+        */
+        stage('System and Environment Tests') { 
+          agent { label 'linux && immutable' }
+          options { skipDefaultCheckout() }
+          environment {
+            PATH = "${env.PATH}:${env.WORKSPACE}/bin"
+            HOME = "${env.WORKSPACE}"
+            GOPATH = "${env.WORKSPACE}"
+          }
+          when { 
+            beforeAgent true
+            environment name: 'test_sys_env_ci', value: 'true' 
+          }
+          steps {
+            withEnvWrapper() {
+              unstash 'source'
+              dir("${BASE_DIR}"){
+                sh './script/jenkins/test.sh'
+              }
+            }
+          }
+          post { 
+            always {
+              coverageReport("${BASE_DIR}/build/coverage")
+              junit(allowEmptyResults: true, 
+                keepLongStdio: true, 
+                testResults: "${BASE_DIR}/build/junit-*.xml,${BASE_DIR}/build/TEST-*.xml")
+              //googleStorageUpload bucket: "gs://${JOB_GCS_BUCKET}/${JOB_NAME}/${BUILD_NUMBER}", credentialsId: "${JOB_GCS_CREDENTIALS}", pathPrefix: "${BASE_DIR}", pattern: '**/build/system-tests/run/**/*', sharedPublicly: true, showInline: true
+              //googleStorageUpload bucket: "gs://${JOB_GCS_BUCKET}/${JOB_NAME}/${BUILD_NUMBER}", credentialsId: "${JOB_GCS_CREDENTIALS}", pathPrefix: "${BASE_DIR}", pattern: '**/build/TEST-*.out', sharedPublicly: true, showInline: true
+              tar(file: "system-tests-linux-files.tgz", archive: true, dir: "system-tests", pathPrefix: "${BASE_DIR}/build")
+              tar(file: "coverage-files.tgz", archive: true, dir: "coverage", pathPrefix: "${BASE_DIR}/build")
+              codecov(repo: 'apm-server', basedir: "${BASE_DIR}")
+            }
+          }
+        }
+        /**
+        Run tests on a windows environment.
+        Finally archive the results.
+        */
+        stage('windows test') { 
```

Comment:
This could be run in parallel with the linux tests.

Reply:
it does

<img width="856" alt="screenshot 2018-12-04 at 13 52 17" src="https://user-images.githubusercontent.com/5400788/49444452-85ce9780-f7cf-11e8-9f5a-293b029bda6c.png">